database.go

```diff
@@ -0,0 +1,723 @@
+	}
+	return shards
+}
+
+// Filter represents a tag filter when looking up other tags or measurements.
+type Filter struct {
+	Not   bool
+	Key   string
+	Value string
+	Regex *regexp.Regexp
+}
+
+type Filters []*Filter
+
+// SeriesIDs is a convenience type for sorting, checking equality, and doing union and
+// intersection of collections of series ids.
+type SeriesIDs []uint32
+
+func (p SeriesIDs) Len() int           { return len(p) }
+func (p SeriesIDs) Less(i, j int) bool { return p[i] < p[j] }
+func (p SeriesIDs) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+// Equals assumes that both are sorted. This is by design, no touchy!
+func (a SeriesIDs) Equals(seriesIDs SeriesIDs) bool {
+	if len(a) != len(seriesIDs) {
+		return false
+	}
+	for i, s := range seriesIDs {
+		if a[i] != s {
+			return false
+		}
+	}
+	return true
+}
+
+// Intersect returns a new collection of series ids in sorted order that is the intersection of the two.
+// The two collections must already be sorted.
+func (a SeriesIDs) Intersect(seriesIDs SeriesIDs) SeriesIDs {
+	l := a
+	r := seriesIDs
+
+	// we want to iterate through the shortest one and stop
+	if len(seriesIDs) < len(a) {
+		l = seriesIDs
+		r = a
+	}
+
+	// they're in sorted order so advance the counter as needed.
+	// That is, don't run comparisons against lower values that we've already passed
+	var i, j int
+
+	ids := make([]uint32, 0, len(l))
+	for i < len(l) {
+		if l[i] == r[j] {
+			ids = append(ids, l[i])
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			i += 1
+		} else {
+			j += 1
+		}
+	}
+
+	return SeriesIDs(ids)
+}
+
+// Union returns a new collection of series ids in sorted order that is the union of the two.
+// The two collections must already be sorted.
+func (l SeriesIDs) Union(r SeriesIDs) SeriesIDs {
+	ids := make([]uint32, 0, len(l)+len(r))
+	var i, j int
+	for i < len(l) && j < len(r) {
+		if l[i] == r[j] {
+			ids = append(ids, l[i])
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			ids = append(ids, l[i])
+			i += 1
+		} else {
+			ids = append(ids, r[j])
+			j += 1
+		}
+	}
+
+	// now append the remainder
+	if i < len(l) {
+		ids = append(ids, l[i:]...)
+	} else if j < len(r) {
+		ids = append(ids, r[j:]...)
+	}
+
+	return ids
+}
+
+// Reject returns a new collection of series ids in sorted order with the passed in set removed from the original. This is useful for the NOT operator.
+// The two collections must already be sorted.
+func (l SeriesIDs) Reject(r SeriesIDs) SeriesIDs {
+	var i, j int
+
+	ids := make([]uint32, 0, len(l))
+	for i < len(l) && j < len(r) {
+		if l[i] == r[j] {
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			ids = append(ids, l[i])
+			i += 1
+		} else {
+			j += 1
+		}
+	}
+
+	// append the remainder
+	if i < len(l) {
+		ids = append(ids, l[i:]...)
+	}
+
+	return SeriesIDs(ids)
+}
+
+// AddSeries adds the series for the given measurement to the index. Returns false if already present
+func (t *database) AddSeries(name string, s *Series) bool {
```

Comment:
If you called this even `m` it would be clearer to me what this variable is. But that may just be me. `name` is pretty general. Also, I think this function would be better named `AddSeriesToIndex`. As it stands it sounds like it's adding a series to the system, which it is not doing -- the series already exists.